src/task_manager.cc

```diff
@@ -181,71 +202,92 @@ namespace detail {
 		max_pseudo_critical_path_length = std::max(max_pseudo_critical_path_length, depender->get_pseudo_critical_path_length());
 	}
 
+	task_id task_manager::collect_execution_front(std::unique_ptr<task> reducer) {
+		// add dependencies from a copy of the front to this task
+		const auto current_front = execution_front;
+		for(task* front_task : current_front) {
+			add_dependency(reducer.get(), front_task, dependency_kind::TRUE_DEP, dependency_origin::execution_front);
+		}
+		assert(execution_front.empty());
+		return register_task_internal(std::move(reducer)).get_id();
+	}
+
+	void task_manager::set_current_epoch(const task_id epoch) {
+		// apply the new epoch to buffers_last_writers and last_collective_tasks data structs
+		for(auto& [_, buffer_region_map] : buffers_last_writers) {
+			buffer_region_map.apply_to_values([epoch](const std::optional<task_id> tid) -> std::optional<task_id> {
+				if(!tid) return tid;
+				return {std::max(epoch, *tid)};
+			});
+		}
+		for(auto& [cgid, tid] : last_collective_tasks) {
+			tid = std::max(epoch, tid);
+		}
+		for(auto& [hoid, tid] : host_object_last_effects) {
+			tid = std::max(epoch, tid);
+		}
+
+		current_epoch = epoch;
+	}
+
 	void task_manager::generate_task_horizon() {
 		// we are probably overzealous in locking here
 		{
 			std::lock_guard lock(task_mutex);
 			current_horizon_critical_path_length = max_pseudo_critical_path_length;
+			const auto previous_horizon = current_horizon;
+			current_horizon = collect_execution_front(task::make_horizon_task(get_new_tid()));
+			if(previous_horizon) { set_current_epoch(*previous_horizon); }
+		}
 
-			auto* previous_horizon_task = current_horizon_task;
-			current_horizon_task = &register_task_internal(task::make_horizon_task(get_new_tid()));
-
-			// add dependencies from a copy of the front to this task
-			auto current_front = get_execution_front();
-			for(task* front_task : current_front) {
-				if(front_task != current_horizon_task) { add_dependency(current_horizon_task, front_task); }
-			}
-
-			// apply the previous horizon to buffers_last_writers and last_collective_tasks data structs
-			if(previous_horizon_task != nullptr) {
-				const task_id prev_hid = previous_horizon_task->get_id();
-				for(auto& [_, buffer_region_map] : buffers_last_writers) {
-					buffer_region_map.apply_to_values([prev_hid](std::optional<task_id> tid) -> std::optional<task_id> {
-						if(!tid) return tid;
-						return {std::max(prev_hid, *tid)};
-					});
-				}
-				for(auto& [cgid, tid] : last_collective_tasks) {
-					tid = std::max(prev_hid, tid);
-				}
-				for(auto& [hoid, tid] : host_object_last_effects) {
-					tid = std::max(prev_hid, tid);
-				}
+		// it's important that we don't hold the lock while doing this
+		invoke_callbacks(*current_horizon);
+	}
 
-				// We also use the previous horizon as the new init task for host-initialized buffers
-				current_init_task_id = prev_hid;
-			}
+	task_id task_manager::finish_epoch(epoch_action action) {
```

Comment:
As per the previous comment, in this context, I'd now prefer `generate_epoch` even more, especially vis-a-vis `generate_horizon`.